dali/core/mm/malloc_resource.cc

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <mutex>
+#include <unordered_map>
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/core/device_guard.h"
+#include "dali/core/spinlock.h"
+#include "dali/core/mm/malloc_resource.h"
+#include "dali/core/mm/detail/align.h"
+
+namespace dali {
+namespace mm {
+
+#if CUDA_VERSION >= 11020
+
+namespace {
+
+class aligned_alloc_helper {
+ public:
+  static constexpr size_t kCudaMallocAlignment = 256;
+
+  void *alloc(size_t size, size_t align, cudaStream_t stream) {
+    void *ptr;
+    if (align > kCudaMallocAlignment) {
+      CUDA_CALL(cudaMallocAsync(&ptr, size + align - kCudaMallocAlignment, stream));
+      if (detail::is_aligned(ptr, size))
+        return ptr;
+
+      void *aligned = detail::align_ptr(ptr, size);
+      std::lock_guard guard(mtx_);
+      orig_ptrs_[aligned] = ptr;;
+      return aligned;
+    } else {
+      CUDA_CALL(cudaMallocAsync(&ptr, size, stream));
+    }
+    return ptr;
+  }
+
+  void free(void *ptr, size_t size, size_t alignment, cudaStream_t stream) {
+    (void)size;
+    if (alignment > kCudaMallocAlignment) {
+      auto it = orig_ptrs_.find(ptr);
+      std::lock_guard guard(mtx_);
+      if (it != orig_ptrs_.end()) {
+        ptr = it->second;
+        orig_ptrs_.erase(it);
+      }
+    }
+    CUDA_DTOR_CALL(cudaFreeAsync(ptr, stream));
+  }
+
+  static aligned_alloc_helper &instance() {
+    static aligned_alloc_helper inst;
+    return inst;
+  }
+
+ private:
+  spinlock mtx_;
+  std::unordered_map<void *, void*> orig_ptrs_;
+};
+
+}  // namespace
+
+cuda_malloc_async_memory_resource::cuda_malloc_async_memory_resource(int device_id) {
+  // Construct the helper.
+  // The `instance` construction is finished before that of the first instance of
```

Comment:
```suggestion
  // The `instance` construction is finished before that of the first instance of
```
I'm not sure if I understand that sentence.

Reply:
I'll rephrase.